Description: Add several video directories to vdr
 and display all recordings in one list
Origin: https://github.com/flensrocker/vdr-plugin-dbus2vdr
Forwarded: no
Author: Lars Hanisch <dvb@flensrocker.de>

--- a/menu.c
+++ b/menu.c
@@ -2706,6 +2706,19 @@
      char *s = ExchangeChars(strdup(base), true);
      d = AddDirectory(d, s);
      free(s);
+     if (!DirectoryOk(*d, false, true)) {
+        cString e;
+        if (LockExtraVideoDirectories(true)) {
+           for (int i = 0; i < ExtraVideoDirectories.Size(); i++) {
+               e = AddDirectory(ExtraVideoDirectories.At(i), s);
+               if (DirectoryOk(*e, false, true)) {
+                  UnlockExtraVideoDirectories();
+                  return e;
+                  }
+               }
+           UnlockExtraVideoDirectories();
+           }
+        }
      }
   return d;
 }
@@ -2736,7 +2749,7 @@
      if (ri->IsDirectory())
         Open();
      else {
-        cReplayControl::SetRecording(ri->Recording()->FileName());
+        cReplayControl::SetRecording(ri->Recording()->FileName(), ri->Recording()->VideoDir());
         return osReplay;
         }
      }
@@ -5165,6 +5178,7 @@
 
 cReplayControl *cReplayControl::currentReplayControl = NULL;
 cString cReplayControl::fileName;
+cString cReplayControl::videoDir;
 
 cReplayControl::cReplayControl(bool PauseLive, bool ReusedPauseLive)
 :cDvbPlayerControl(fileName, PauseLive, ReusedPauseLive)
@@ -5182,7 +5196,7 @@
   lastSkipTimeout.Set(0);
   timeoutShow = 0;
   timeSearchActive = false;
-  cRecording Recording(fileName);
+  cRecording Recording(fileName, videoDir);
   cStatus::MsgReplaying(this, Recording.Name(), Recording.FileName(), true);
   marks.Load(fileName, Recording.FramesPerSecond(), Recording.IsPesRecording());
   SetTrackDescriptions(false);
@@ -5232,9 +5246,12 @@
   cDvbPlayerControl::Stop();
 }
 
-void cReplayControl::SetRecording(const char *FileName)
+void cReplayControl::SetRecording(const char *FileName, const char *VideoDir)
 {
   fileName = FileName;
+  videoDir = VideoDir;
+  if ((FileName != NULL) && (VideoDir == NULL))
+     videoDir = FindMatchingExtraVideoDirectory(FileName);
 }
 
 const char *cReplayControl::NowReplaying(void)
--- a/menu.h
+++ b/menu.h
@@ -292,6 +292,7 @@
   void ShowTimed(int Seconds = 0);
   static cReplayControl *currentReplayControl;
   static cString fileName;
+  static cString videoDir;
   void ShowMode(void);
   bool ShowProgress(bool Initial);
   void MarkToggle(void);
@@ -309,7 +310,7 @@
   virtual void Show(void);
   virtual void Hide(void);
   bool Visible(void) { return visible; }
-  static void SetRecording(const char *FileName);
+  static void SetRecording(const char *FileName, const char *VideoDir = NULL);
   static const char *NowReplaying(void);
   static const char *LastReplayed(void);
   static void ClearLastReplayed(const char *FileName);
--- a/recording.c
+++ b/recording.c
@@ -740,6 +740,7 @@
   resume = RESUME_NOT_INITIALIZED;
   titleBuffer = NULL;
   sortBufferName = sortBufferTime = NULL;
+  videoDir = VideoDirectory;
   fileName = NULL;
   name = NULL;
   fileSizeMB = -1; // unknown
@@ -792,7 +793,7 @@
   info->lifetime = lifetime;
 }
 
-cRecording::cRecording(const char *FileName)
+cRecording::cRecording(const char *FileName, const char *VideoDir)
 {
   resume = RESUME_NOT_INITIALIZED;
   fileSizeMB = -1; // unknown
@@ -807,11 +808,20 @@
   deleted = 0;
   titleBuffer = NULL;
   sortBufferName = sortBufferTime = NULL;
+  if (VideoDir == NULL) {
+     cString extraVideoDir = FindMatchingExtraVideoDirectory(FileName);
+     if (*extraVideoDir == NULL)
+        videoDir = VideoDirectory;
+     else
+        videoDir = strdup(*extraVideoDir);
+     }
+  else
+     videoDir = strdup(VideoDir);
   FileName = fileName = strdup(FileName);
   if (*(fileName + strlen(fileName) - 1) == '/')
      *(fileName + strlen(fileName) - 1) = 0;
-  if (strstr(FileName, VideoDirectory) == FileName)
-     FileName += strlen(VideoDirectory) + 1;
+  if (strstr(FileName, videoDir) == FileName)
+     FileName += strlen(videoDir) + 1;
   const char *p = strrchr(FileName, '/');
 
   name = NULL;
@@ -918,6 +928,8 @@
 
 cRecording::~cRecording()
 {
+  if (videoDir != VideoDirectory)
+     free((char*)videoDir);
   free(titleBuffer);
   free(sortBufferName);
   free(sortBufferTime);
@@ -966,7 +978,7 @@
         *sb = strdup(buf);
         }
      else {
-        char *s = strdup(FileName() + strlen(VideoDirectory));
+        char *s = strdup(FileName() + strlen(videoDir));
         if (RecordingsSortMode != rsmName || Setup.AlwaysSortFoldersFirst)
            s = StripEpisodeName(s, RecordingsSortMode != rsmName);
         strreplace(s, '/', '0'); // some locales ignore '/' when sorting
@@ -1008,11 +1020,11 @@
      const char *fmt = isPesRecording ? NAMEFORMATPES : NAMEFORMATTS;
      int ch = isPesRecording ? priority : channel;
      int ri = isPesRecording ? lifetime : instanceId;
-     char *Name = LimitNameLengths(strdup(name), DirectoryPathMax - strlen(VideoDirectory) - 1 - 42, DirectoryNameMax); // 42 = length of an actual recording directory name (generated with DATAFORMATTS) plus some reserve
+     char *Name = LimitNameLengths(strdup(name), DirectoryPathMax - strlen(videoDir) - 1 - 42, DirectoryNameMax); // 42 = length of an actual recording directory name (generated with DATAFORMATTS) plus some reserve
      if (strcmp(Name, name) != 0)
         dsyslog("recording file name '%s' truncated to '%s'", name, Name);
      Name = ExchangeChars(Name, true);
-     fileName = strdup(cString::sprintf(fmt, VideoDirectory, Name, t->tm_year + 1900, t->tm_mon + 1, t->tm_mday, t->tm_hour, t->tm_min, ch, ri));
+     fileName = strdup(cString::sprintf(fmt, videoDir, Name, t->tm_year + 1900, t->tm_mon + 1, t->tm_mday, t->tm_hour, t->tm_min, ch, ri));
      free(Name);
      }
   return fileName;
@@ -1291,9 +1303,14 @@
   ChangeState();
   Unlock();
   ScanVideoDir(VideoDirectory, Foreground);
+  if (LockExtraVideoDirectories()) {
+     for (int i = 0; i < ExtraVideoDirectories.Size(); i++)
+         ScanVideoDir(ExtraVideoDirectories.At(i), Foreground, ExtraVideoDirectories.At(i));
+     UnlockExtraVideoDirectories();
+     }
 }
 
-void cRecordings::ScanVideoDir(const char *DirName, bool Foreground, int LinkLevel, int DirLevel)
+void cRecordings::ScanVideoDir(const char *DirName, bool Foreground, int LinkLevel, const char *BaseVideoDir)
 {
   // Find any new recordings:
   cReadDir d(DirName);
@@ -1315,7 +1332,7 @@
            if (S_ISDIR(st.st_mode)) {
               if (endswith(buffer, deleted ? DELEXT : RECEXT)) {
                  if (deleted || initial || !GetByName(buffer)) {
-                    cRecording *r = new cRecording(buffer);
+                    cRecording *r = new cRecording(buffer, BaseVideoDir);
                     if (r->Name()) {
                        r->NumFrames(); // initializes the numFrames member
                        r->FileSizeMB(); // initializes the fileSizeMB member
@@ -1332,7 +1349,7 @@
                     }
                  }
               else
-                 ScanVideoDir(buffer, Foreground, LinkLevel + Link, DirLevel + 1);
+                 ScanVideoDir(buffer, Foreground, LinkLevel + Link, BaseVideoDir);
               }
            }
         }
--- a/recording.h
+++ b/recording.h
@@ -107,9 +107,10 @@
   int priority;
   int lifetime;
   time_t deleted;
+  const char *videoDir;
 public:
   cRecording(cTimer *Timer, const cEvent *Event);
-  cRecording(const char *FileName);
+  cRecording(const char *FileName, const char *VideoDir = NULL);
   virtual ~cRecording();
   time_t Start(void) const { return start; }
   int Priority(void) const { return priority; }
@@ -118,6 +119,7 @@
   virtual int Compare(const cListObject &ListObject) const;
   const char *Name(void) const { return name; }
   const char *FileName(void) const;
+  const char *VideoDir(void) const { return videoDir; }
   const char *Title(char Delimiter = ' ', bool NewIndicator = false, int Level = -1) const;
   const cRecordingInfo *Info(void) const { return info; }
   const char *PrefixFileName(char Prefix);
@@ -168,7 +170,7 @@
   int state;
   const char *UpdateFileName(void);
   void Refresh(bool Foreground = false);
-  void ScanVideoDir(const char *DirName, bool Foreground = false, int LinkLevel = 0, int DirLevel = 0);
+  void ScanVideoDir(const char *DirName, bool Foreground = false, int LinkLevel = 0, const char *BaseVideoDir = NULL);
 protected:
   void Action(void);
 public:
--- a/svdrp.c
+++ b/svdrp.c
@@ -334,6 +334,14 @@
   "    be turned up or down, respectively. The option 'mute' will toggle the\n"
   "    audio muting. If no option is given, the current audio volume level will\n"
   "    be returned.",
+  "AXVD directory\n"
+  "    add directory to extra video directory list",
+  "CXVD\n"
+  "    clear extra video directory list",
+  "DXVD directory\n"
+  "    delete directory from extra video directory list",
+  "LXVD\n"
+  "    list extra video directories",
   "QUIT\n"
   "    Exit vdr (SVDRP).\n"
   "    You can also hit Ctrl-D to exit.",
@@ -1472,7 +1480,7 @@
               else
                  resume.Save(pos);
               }
-           cReplayControl::SetRecording(recording->FileName());
+           cReplayControl::SetRecording(recording->FileName(), recording->VideoDir());
            cControl::Launch(new cReplayControl);
            cControl::Attach();
            Reply(250, "Playing recording \"%s\" [%s]", num, recording->Title());
@@ -1683,6 +1691,62 @@
      Reply(250, "Audio volume is %d", cDevice::CurrentVolume());
 }
 
+void cSVDRP::CmdAXVD(const char *Option)
+{
+  if (*Option) {
+     if (!LockExtraVideoDirectories(false)) {
+        Reply(550, "Unable to lock extra video directory list");
+        return;
+        }
+     AddExtraVideoDirectory(Option);
+     UnlockExtraVideoDirectories();
+     Reply(250, "added '%s' to extra video directory list", Option);
+     return;
+     }
+  Reply(501, "Missing directory name");
+}
+
+void cSVDRP::CmdCXVD(const char *Option)
+{
+  if (!LockExtraVideoDirectories(false)) {
+     Reply(550, "Unable to lock extra video directory list");
+     return;
+     }
+  ExtraVideoDirectories.Clear();
+  UnlockExtraVideoDirectories();
+  Reply(250, "cleared extra video directory list");
+}
+
+void cSVDRP::CmdDXVD(const char *Option)
+{
+  if (*Option) {
+     if (!LockExtraVideoDirectories(false)) {
+        Reply(550, "Unable to lock extra video directory list");
+        return;
+        }
+     DelExtraVideoDirectory(Option);
+     UnlockExtraVideoDirectories();
+     Reply(250, "removed '%s' from extra video directory list", Option);
+     return;
+     }
+  Reply(501, "Missing directory name");
+}
+
+void cSVDRP::CmdLXVD(const char *Option)
+{
+  if (!LockExtraVideoDirectories(false)) {
+     Reply(550, "Unable to lock extra video directory list");
+     return;
+     }
+  if (ExtraVideoDirectories.Size() == 0)
+     Reply(550, "no extra video directories in list");
+  else {
+     for (int i = 0; i < ExtraVideoDirectories.Size(); i++)
+         Reply(i < ExtraVideoDirectories.Size() - 1 ? -250 : 250, "%s", ExtraVideoDirectories.At(i));
+     }
+  UnlockExtraVideoDirectories();
+}
+
 #define CMD(c) (strcasecmp(Cmd, c) == 0)
 
 void cSVDRP::Execute(char *Cmd)
@@ -1736,6 +1800,10 @@
   else if (CMD("UPDR"))  CmdUPDR(s);
   else if (CMD("UPDT"))  CmdUPDT(s);
   else if (CMD("VOLU"))  CmdVOLU(s);
+  else if (CMD("AXVD"))  CmdAXVD(s);
+  else if (CMD("CXVD"))  CmdCXVD(s);
+  else if (CMD("DXVD"))  CmdDXVD(s);
+  else if (CMD("LXVD"))  CmdLXVD(s);
   else if (CMD("QUIT"))  Close(true);
   else                   Reply(500, "Command unrecognized: \"%s\"", Cmd);
 }
--- a/svdrp.h
+++ b/svdrp.h
@@ -85,6 +85,10 @@
   void CmdUPDT(const char *Option);
   void CmdUPDR(const char *Option);
   void CmdVOLU(const char *Option);
+  void CmdAXVD(const char *Option);
+  void CmdCXVD(const char *Option);
+  void CmdDXVD(const char *Option);
+  void CmdLXVD(const char *Option);
   void Execute(char *Cmd);
 public:
   cSVDRP(int Port);
--- a/tools.c
+++ b/tools.c
@@ -375,12 +375,15 @@
   return Free;
 }
 
-bool DirectoryOk(const char *DirName, bool LogErrors)
+bool DirectoryOk(const char *DirName, bool LogErrors, bool JustReadOnly)
 {
+  int mode = R_OK;
+  if (!JustReadOnly)
+     mode |= W_OK | X_OK;
   struct stat ds;
   if (stat(DirName, &ds) == 0) {
      if (S_ISDIR(ds.st_mode)) {
-        if (access(DirName, R_OK | W_OK | X_OK) == 0)
+        if (access(DirName, mode) == 0)
            return true;
         else if (LogErrors)
            esyslog("ERROR: can't access %s", DirName);
--- a/tools.h
+++ b/tools.h
@@ -228,7 +228,7 @@
 cString AddDirectory(const char *DirName, const char *FileName);
 bool EntriesOnSameFileSystem(const char *File1, const char *File2);
 int FreeDiskSpaceMB(const char *Directory, int *UsedMB = NULL);
-bool DirectoryOk(const char *DirName, bool LogErrors = false);
+bool DirectoryOk(const char *DirName, bool LogErrors = false, bool JustReadOnly = false);
 bool MakeDirs(const char *FileName, bool IsDirectory = false);
 bool RemoveFileOrDir(const char *FileName, bool FollowSymlinks = false);
 bool RemoveEmptyDirectories(const char *DirName, bool RemoveThis = false, const char *IgnoreFiles[] = NULL);
--- a/vdr.c
+++ b/vdr.c
@@ -200,6 +200,7 @@
   int SVDRPport = DEFAULTSVDRPPORT;
   const char *AudioCommand = NULL;
   const char *VideoDirectory = DEFAULTVIDEODIR;
+  const char *ExtraVideoDirectory = NULL;
   const char *ConfigDirectory = NULL;
   const char *CacheDirectory = NULL;
   const char *ResourceDirectory = NULL;
@@ -263,6 +264,7 @@
       { "version",  no_argument,       NULL, 'V' },
       { "vfat",     no_argument,       NULL, 'v' | 0x100 },
       { "video",    required_argument, NULL, 'v' },
+      { "extravideo", required_argument, NULL, 'v' | 0x200 },
       { "watchdog", required_argument, NULL, 'w' },
       { NULL,       no_argument,       NULL,  0  }
     };
@@ -455,6 +457,12 @@
                           optarg[strlen(optarg) - 1] = 0;
                     SetVideoDirectory(VideoDirectory);
                     break;
+          case 'v' | 0x200:
+                    ExtraVideoDirectory = optarg;
+                    while (optarg && *optarg && optarg[strlen(optarg) - 1] == '/')
+                          optarg[strlen(optarg) - 1] = 0;
+                    AddExtraVideoDirectory(ExtraVideoDirectory);
+                    break;
           case 'w': if (isnumber(optarg)) {
                        int t = atoi(optarg);
                        if (t >= 0) {
@@ -552,6 +560,8 @@
                "                           root\n"
                "            --userdump     allow coredumps if -u is given (debugging)\n"
                "  -v DIR,   --video=DIR    use DIR as video directory (default: %s)\n"
+               "            --extravideo=DIR use DIR as an additional readonly video directory\n"
+               "                           can be used multiple times\n"
                "  -V,       --version      print version information and exit\n"
                "            --vfat         for backwards compatibility (same as\n"
                "                           --dirnames=250,40,1)\n"
--- a/videodir.c
+++ b/videodir.c
@@ -19,6 +19,65 @@
 #include "recording.h"
 #include "tools.h"
 
+cStringList ExtraVideoDirectories;
+bool        ExtraVideoDirectoriesIsLocked = false;
+cMutex      ExtraVideoDirectoriesMutex;
+
+bool LockExtraVideoDirectories(bool Wait)
+{
+  if (!Wait && ExtraVideoDirectoriesIsLocked)
+     return false;
+  ExtraVideoDirectoriesMutex.Lock();
+  ExtraVideoDirectoriesIsLocked = true;
+  return true;
+}
+
+void UnlockExtraVideoDirectories(void)
+{
+  ExtraVideoDirectoriesIsLocked = false;
+  ExtraVideoDirectoriesMutex.Unlock();
+}
+
+void AddExtraVideoDirectory(const char *Directory)
+{
+  if ((Directory != NULL) && (ExtraVideoDirectories.Find(Directory) < 0))
+     ExtraVideoDirectories.Append(strdup(Directory));
+}
+
+void DelExtraVideoDirectory(const char *Directory)
+{
+  if (Directory != NULL) {
+     int index = ExtraVideoDirectories.Find(Directory);
+     if (index < 0)
+        return;
+     char *dir = ExtraVideoDirectories.At(index);
+     ExtraVideoDirectories.Remove(index);
+     free(dir);
+     }
+}
+
+cString FindMatchingExtraVideoDirectory(const char *FileName)
+{
+  if (FileName == NULL)
+     return cString(NULL);
+  uint fileLen = strlen(FileName);
+  if ((strlen(VideoDirectory) < fileLen) && startswith(FileName, VideoDirectory))
+     return cString(NULL);
+  if (!LockExtraVideoDirectories())
+     return cString(NULL);
+  int i = 0;
+  cString videoDir(NULL);
+  while (i < ExtraVideoDirectories.Size()) {
+        if ((strlen(ExtraVideoDirectories.At(i)) < fileLen) && startswith(FileName, ExtraVideoDirectories.At(i))) {
+           videoDir = ExtraVideoDirectories.At(i);
+           break;
+           }
+        i++;
+        }
+  UnlockExtraVideoDirectories();
+  return videoDir;
+}
+
 const char *VideoDirectory = VIDEODIR;
 
 void SetVideoDirectory(const char *Directory)
@@ -28,13 +87,13 @@
 
 class cVideoDirectory {
 private:
-  char *name, *stored, *adjusted;
+  char *videoDir, *name, *stored, *adjusted;
   int length, number, digits;
 public:
-  cVideoDirectory(void);
+  cVideoDirectory(const char *VideoDir = NULL);
   ~cVideoDirectory();
   int FreeMB(int *UsedMB = NULL);
-  const char *Name(void) { return name ? name : VideoDirectory; }
+  const char *Name(void) { return name ? name : videoDir; }
   const char *Stored(void) { return stored; }
   int Length(void) { return length; }
   bool IsDistributed(void) { return name != NULL; }
@@ -43,10 +102,14 @@
   const char *Adjust(const char *FileName);
   };
 
-cVideoDirectory::cVideoDirectory(void)
+cVideoDirectory::cVideoDirectory(const char *VideoDir)
 {
-  length = strlen(VideoDirectory);
-  name = (VideoDirectory[length - 1] == '0') ? strdup(VideoDirectory) : NULL;
+  if (VideoDir == NULL)
+     videoDir = strdup(VideoDirectory);
+  else
+     videoDir = strdup(VideoDir);
+  length = strlen(videoDir);
+  name = (videoDir[length - 1] == '0') ? strdup(videoDir) : NULL;
   stored = adjusted = NULL;
   number = -1;
   digits = 0;
@@ -54,6 +117,7 @@
 
 cVideoDirectory::~cVideoDirectory()
 {
+  free(videoDir);
   free(name);
   free(stored);
   free(adjusted);
@@ -61,7 +125,7 @@
 
 int cVideoDirectory::FreeMB(int *UsedMB)
 {
-  return FreeDiskSpaceMB(name ? name : VideoDirectory, UsedMB);
+  return FreeDiskSpaceMB(name ? name : videoDir, UsedMB);
 }
 
 bool cVideoDirectory::Next(void)
@@ -113,14 +177,18 @@
   const char *ActualFileName = FileName;
 
   // Incoming name must be in base video directory:
+  cString extraVideoDir;
   if (strstr(FileName, VideoDirectory) != FileName) {
-     esyslog("ERROR: %s not in %s", FileName, VideoDirectory);
-     errno = ENOENT; // must set 'errno' - any ideas for a better value?
-     return NULL;
+     extraVideoDir = FindMatchingExtraVideoDirectory(FileName);
+     if (*extraVideoDir == NULL) {
+        esyslog("ERROR: %s not in %s", FileName, VideoDirectory);
+        errno = ENOENT; // must set 'errno' - any ideas for a better value?
+        return NULL;
+        }
      }
   // Are we going to create a new file?
   if ((Flags & O_CREAT) != 0) {
-     cVideoDirectory Dir;
+     cVideoDirectory Dir(*extraVideoDir);
      if (Dir.IsDistributed()) {
         // Find the directory with the most free space:
         int MaxFree = Dir.FreeMB();
--- a/videodir.h
+++ b/videodir.h
@@ -13,6 +13,16 @@
 #include <stdlib.h>
 #include "tools.h"
 
+#define EXTRA_VIDEO_DIRECTORIES_PATCH 1
+
+extern cStringList ExtraVideoDirectories;
+
+bool LockExtraVideoDirectories(bool Wait = true);
+void UnlockExtraVideoDirectories(void);
+void AddExtraVideoDirectory(const char *Directory);
+void DelExtraVideoDirectory(const char *Directory);
+cString FindMatchingExtraVideoDirectory(const char *FileName);
+
 extern const char *VideoDirectory;
 
 void SetVideoDirectory(const char *Directory);
